### PR TITLE
bsc#1169410 fixes

### PR DESCRIPTION
**Pull request by imobachgs**
This PR fixes several issues described in [bsc#1169410](https://bugzilla.suse.com/show_bug.cgi?id=1169410). It is the continuation of #80.

- Add an option to set the log level when running the provisioners. By default, it is set to "info".
- By default, do not retry in masterless mode.
- Honor the AutoYaST reporting settings.
- When not running during autoinstallation, do not automatically close the dialog after finishing.
- Allow using several locations to save pillars data.
- Do not leak passwords from pillars to the console.
- Add the `/usr/share/salt-formulas` to the list of directories to search for formulas.
- Groups and namespaces are presented in the same way due to some UI limitations.
- Fixes several problems related to widgets visibility.
- Take precedence over existing top.sls files.

Trello: https://trello.com/c/S7XSVGWs/

visibility

### CODE REVIEW

File: src/lib/y2configuration_management/dialogs/running.rb [hunk 1]
```diff
@@ -15,7 +15,27 @@ class Running < ::UI::Dialog
...
+      # @param reporting_opts [Hash] Reporting reporting_opts
+      # @option reporting_opts [Boolean] :open_after_success Keep the dialog open after finishing
+      #   successfuly
+      # @option reporting_opts [Boolean] :open_after_error Keep the dialog open after finishing
+      #   with an error
+      # @option reporting_opts [Integer] :timeout_after_success Timeout after finishing successfuly.
+      #   It only makes sense when `:open_after_success` is set to true.
+      # @option reporting_opts [Integer] :timeout_after_error Timeout after finishing with an error
+      #   It only makes sense when `:open_after_error` is set to true.
+      def initialize(reporting_opts: {})
```
Comment (jreidinger) [inline src/lib/y2configuration_management/dialogs/running.rb:31]: hmm, in general this class looks pretty generic and not specific to configuration management. Don't we already have something like this? And if not, then it makes sense for me to have it in yast2 itself.

File: src/lib/y2configuration_management/salt/form_data_writer.rb [hunk 2]
```diff
@@ -58,12 +58,11 @@ def to_pillar_data
...
-        case data
-        when Array
+        if data.is_a?(Array)
           collection_for_pillar(data, element)
-        when Hash
+        elsif data.is_a?(Hash)
           hash_for_pillar(data, element)
-        else
+        elsif element.is_a?(FormInput)
           scalar_for_pillar(data, element)
```
Comment (jreidinger) [inline src/lib/y2configuration_management/salt/form_data_writer.rb:66]: what about remaining else? and maybe when you use that if cascade, I would integrate also nil case which is now on top.

File: src/lib/y2configuration_management/widgets/pager_tree_item.rb [hunk 3]
```diff
@@ -33,16 +33,25 @@ class PagerTreeItem < ::CWM::PagerTreeItem
...
       # @param page [Page] Associated page
       # @param icon
-      def initialize(page, icon: nil, open: true, children: [])
-        super
+      # rubocop:disable Metrics/ParameterLists
+      def initialize(page, icon: nil, open: true, children: [], locator: nil, visible_if: nil)
+        super(page, icon: icon, open: open, children: children)
+        @locator = locator
+        @visible_if = visible_if
+        @visible = true
```
Comment (jreidinger) [inline src/lib/y2configuration_management/widgets/pager_tree_item.rb:50]: new params are not documented

File: src/lib/y2configuration_management/widgets/pager_tree_item.rb [hunk 4]
```diff
@@ -53,6 +62,31 @@ def page_id
...
+      # Updates item visibility
+      #
+      # @param data [FormData]
+      # @see #visible?
+      def update_visibility(data)
+        @visible =
+          if parent && !parent.visible?
+            false
+          elsif visible_if
+            visible_if.evaluate(data, context: self)
```
Comment (jreidinger) [inline src/lib/y2configuration_management/widgets/pager_tree_item.rb:74]: there is no documentation what that visible if is, but why it is not common Proc?

Comment (imobachgs) [reply src/lib/y2configuration_management/widgets/pager_tree_item.rb:74]: Well, visible_if is a [FormCondition](https://github.com/yast/yast-configuration-management/blob/23c85adef2e83859cf4752ae24d38c9756b801bf/src/lib/y2configuration_management/salt/form_condition.rb#L26) instance. Those conditions are used to decide whether an element in the UI is visible or not (as documented in [SUSE Manager Best Practices](https://documentation.suse.com/external-tree/en-us/suma/3.2/susemanager-best-practices/single-html/book.suma.best.practices/book.suma.best.practices.html)).

File: src/lib/y2configuration_management/dialogs/running.rb (hunk 1 above)
Comment (imobachgs) [reply src/lib/y2configuration_management/dialogs/running.rb:31]: That is a fair point. I was tempted to write a new class or even to use the output of `Report.Export`. Finally, I have decided to [create a card](https://trello.com/c/Tpvrm1BV/70-extract-reporting-settings-to-a-separate-class).

**Comment by imobachgs** [comment]
@jreidinger I have done some fixes over the weekend (bascially related to the installation) but one problem still remains and I do not know how to fix it properly. In 2017, we moved the [configuration management](https://github.com/yast/yast-installation/pull/551) to the 1st stage. When running a configuration that contains a SUSE Manager Salt Formula, AutoYaST shows the formulas selection dialog (which is part of [FormulaSequence](https://github.com/yast/yast-configuration-management/blob/master/src/lib/y2configuration_management/salt/formula_sequence.rb)).

formulas-selection

I guess that this dialog replaces the inst_finish UI, so after the `configuration_management_finish` client ends, the same dialog remains and you do not see what YaST is really doing (installing the bootloader).

Is it possible to open the sequence in a different dialog instead of replacing the inst_finish one? A workaround could be to move the `configuration_management_finish` client after the bootloader installation, but I would rather prefer a proper fix.

Finally, I must admit that it is a corner case: if you want to use SUSE Manager Formulas you can run them with yast2-firstboot, which makes more sense. But anyway, I would like to fix it.

**Pull request merged**